### PR TITLE
Create a NETCore-targeted Csc package

### PR DESCRIPTION
**Pull request by agocke**
Create a NETCore-targeted Csc package

Creates a new package, Microsoft.NETCore.Csc, which contains a Csc
binary compiled against the netcoreapp1.1 shared framework. This can be
installed as a NuGet package and built using the dotnet SDK, as long as
the netcoreapp1.1 framework is installed.

This change also downgrades a few packages, like
System.Diagnostics.FileVersionInfo, that were higher than the
version available in netcoreapp1.1.

### CODE REVIEW

**Comment by agocke** [comment]
@jaredpar @dotnet/roslyn-compiler This is ready for review

**Comment by agocke** [comment]
cc @VSadov

File: build/NuGetAdditionalFiles/RunCsc [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+
+THISDIR=$(dirname $0)
+
+dotnet $THISDIR/csc.exe "$@"
```
Comment (sharwell) [inline build/NuGetAdditionalFiles/RunCsc]: :memo: Non-Windows machines are the places where missing trailing newlines caused problems for some cases in the past. While the other cases of this in the PR are 🙁 for me, this one stands out as a possible problem.

File: src/Interactive/CsiCore/project.json [hunk 2]
```diff
@@ -54,7 +54,6 @@
     "System.Xml.XPath.XDocument": "4.3.0"
   },
   "frameworks": {
-    // We don't actually target netstandard1.6; this is to work around https://github.com/dotnet/roslyn/issues/12458
```
Comment (sharwell) [inline src/Interactive/CsiCore/project.json]: ❔ Did you mean to remove this comment? The referenced issue is not marked as resolved.

:memo: This applies to one other file as well.

Comment (agocke) [inline src/Interactive/CsiCore/project.json]: Comments are illegal in JSON.

File: build/NuGetAdditionalFiles/RunCsc [hunk 3]
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
```
Comment (agocke) [inline build/NuGetAdditionalFiles/RunCsc]: This file should be temporary. Also, does it matter what it's called? It won't be on the path.

File: src/Interactive/CsiCore/project.json (hunk 2 above)
Comment (sharwell) [inline src/Interactive/CsiCore/project.json]: @agocke We are not using a strict JSON parser; AFAIK the one we are using does allow comments.

Comment (agocke) [inline src/Interactive/CsiCore/project.json]: We aren't, but someone else may.

Comment (jasonmalinowski) [inline src/Interactive/CsiCore/project.json]: Please leave the comment. Anybody parsing project.json with a strict parser won't parse reality anyways. (That ship sailed along time ago...)

Comment (jasonmalinowski) [inline src/Interactive/CsiCore/project.json]: (it's also possible the bug is fixed, the referenced bug implies so...)

File: build/NuGetAdditionalFiles/RunCsc.cmd [hunk 4]
```diff
@@ -0,0 +1,2 @@
+echo off
+dotnet %~dp0\csc.exe %*
```
Comment (jasonmalinowski) [inline build/NuGetAdditionalFiles/RunCsc.cmd]: dotnet will treat all arguments after the first as arguments to the program?

File: src/NuGet/BuildNuGets.csx [hunk 5]
```diff
@@ -110,6 +116,8 @@ string[] NonRedistPackageNames = {
     "Microsoft.Net.Compilers",
     "Microsoft.Net.Compilers.netcore",
     "Microsoft.Net.CSharp.Interactive.netcore",
+    "Microsoft.NETCore.Csc",
```
Comment (jasonmalinowski) [inline src/NuGet/BuildNuGets.csx]: What's the naming mean here vs. the Microsoft.Net.Compilers.netcore packages?

File: src/NuGet/BuildNuGets.csx [hunk 6]
```diff
@@ -77,6 +77,12 @@ var IsCoreBuild = File.Exists(Path.Combine(ToolsetPath, "corerun"));
 var NuGetAdditionalFilesPath = Path.Combine(SolutionRoot, "build/NuGetAdditionalFiles");
 var ThirdPartyNoticesPath = Path.Combine(NuGetAdditionalFilesPath, "ThirdPartyNotices.rtf");
 var NetCompilersPropsPath = Path.Combine(NuGetAdditionalFilesPath, "Microsoft.Net.Compilers.props");
+var NetcoreCscPropsPath = Path.Combine(NuGetAdditionalFilesPath, "Microsoft.NETCore.Csc.props");
```
Comment (jasonmalinowski) [inline src/NuGet/BuildNuGets.csx]: Should we just pass a single variable "NuGetAdditionalFiles" to the nuget pack invocation at this point?

File: build/NuGetAdditionalFiles/RunCsc (hunk 3 above)
Comment (jasonmalinowski) [inline build/NuGetAdditionalFiles/RunCsc]: I can imagine people in about 5 seconds taking a dependency on the existence of this script in the package. "Oh, I downloaded the package, I guess I run it via running this script..."

File: .gitattributes [hunk 7]
```diff
@@ -12,6 +12,7 @@
 # files must be hard-coded to be cloned with LF
 src/Compilers/CSharp/CscCore/csc text eol=lf
 src/Compilers/VisualBasic/VbcCore/vbc text eol=lf
+build/NuGetAdditionalFiles/RunCsc text eol=lf
```
Comment (jasonmalinowski) [inline .gitattributes]: Why not RunVbc?

File: src/Interactive/CsiCore/project.json (hunk 2 above)
Comment (agocke) [inline src/Interactive/CsiCore/project.json]: @jasonmalinowski I believe it is in the quantum state of both fixed and unfixed. It is fixed in a version of NuGet we do not currently use.

File: build/NuGetAdditionalFiles/RunCsc.cmd (hunk 4 above)
Comment (agocke) [inline build/NuGetAdditionalFiles/RunCsc.cmd]: Yes

File: src/NuGet/BuildNuGets.csx (hunk 6 above)
Comment (agocke) [inline src/NuGet/BuildNuGets.csx]: 🤷‍♂️

File: build/NuGetAdditionalFiles/RunCsc (hunk 3 above)
Comment (agocke) [inline build/NuGetAdditionalFiles/RunCsc]: Once again: why does that matter to me? If you're doing something unsupported why should I give you a friendly name to do that easier?

File: src/NuGet/BuildNuGets.csx (hunk 5 above)
Comment (agocke) [inline src/NuGet/BuildNuGets.csx]: Nothing -- that was a terrible name that I chose at random to give .NET CLI something to use. I'd prefer we forget that package exists.

File: src/Interactive/CsiCore/project.json (hunk 2 above)
Comment (agocke) [inline src/Interactive/CsiCore/project.json]: Also, note that this file is autogenerated by @jaredpar's `RepoUtil`. How would you expect comments in autogenerated files to be preserved?

File: .gitattributes (hunk 7 above)
Comment (agocke) [inline .gitattributes]: Forgot. Nice catch.

File: src/NuGet/BuildNuGets.csx (hunk 6 above)
Comment (jasonmalinowski) [inline src/NuGet/BuildNuGets.csx]: I'll treat 🤷‍♂️ as a yes then.

File: build/NuGetAdditionalFiles/RunCsc (hunk 3 above)
Comment (jasonmalinowski) [inline build/NuGetAdditionalFiles/RunCsc]: I'm questioning your "this should be temporary" statement. :smile:

**Comment by agocke** [comment]
Note: I've unified Microsoft.NETCore.Csc and Micrososft.NETCore.Vbc into Microsoft.NETCore.Compilers. Turns out using both of those in the same project would be impossible because we would be forced to reload the task, so I'm just combining them for now. If we want to make two packages we would have to split the Csc and Vbc tasks into different assemblies.

For reviews: don't worry about the content of the package. I'm more focused on getting this package to users and iterating than trying to get it perfect on the first try.

File: build/NuGetAdditionalFiles/RunCsc (hunk 3 above)
Comment (agocke) [inline build/NuGetAdditionalFiles/RunCsc]: Oh, I'm not worried about breaking people. This package will only end up on MyGet right now.

**Comment by agocke** [comment]
@jasonmalinowski I'm specifically including you due to the change to downgrade some of our packages. Note that those packages are also included as facades in PrivateAssemblies and flow throughout the Roslyn package graph.

File: src/Interactive/CsiCore/project.json (hunk 2 above)
Comment (agocke) [inline src/Interactive/CsiCore/project.json]: I've removed the lines and nothing broke, so :shipit:

Comment (jasonmalinowski) [inline src/Interactive/CsiCore/project.json]: @agocke Are there any additional warning or errors in the Visual Studio error list?

Comment (agocke) [inline src/Interactive/CsiCore/project.json]: @jasonmalinowski Nope. In fact, it removes errors. This construction is illegal in the 4.0 version of NuGet used by VS.

Comment (jasonmalinowski) [inline src/Interactive/CsiCore/project.json]: :shipit:

**Comment by jasonmalinowski** [comment]
@agocke: got it. I still have no clue what the fallout would be. Who originally did the upgrade?

**Comment by agocke** [comment]
@jasonmalinowski Me, I think. I did the wrong version alignment -- I thought 4.3.0 was the version for everything in netcoreapp 1.1 since almost everything was version 4.3.0. In reality, there's no connection. The versioning was just coincidence. There's nothing in the newer versions of these packages that we need.

**Comment by jasonmalinowski** [comment]
> There's nothing in the newer versions of these packages that we need.

OK, that was the statement somebody here needed to make. :smile:

File: build/NuGetAdditionalFiles/Microsoft.NETCore.Compilers.props [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    <VbcToolExe Condition="'$(OS)' != 'Windows_NT'">RunVbc</VbcToolExe>
+  </PropertyGroup>
+
+  <!-- If packaged on Windows, the RunCsc and RunVbc scripts may not be marked
+       executable. If we're not running on Windows, mark them as such, just in case -->
+  <Target Name="MakeCompilerScriptsExecutable"
+          Condition="'$(BuildingProject)' == 'true' AND '$(OS)' != 'Windows_NT'">
+    <Exec Command="chmod +x &quot;$(CscToolPath)/$(CscToolExe)&quot; &quot;$(VbcToolPath)/$(VbcToolExe)&quot;" />
+  </Target>
+</Project>
```
Comment (jaredpar) [inline build/NuGetAdditionalFiles/Microsoft.NETCore.Compilers.props]: This directory name seems pretty odd.  It's not meant for general purpose NuGetAdditionalFiles, it's assets for our NuGet compiler toolset.  Feel like this directory is misnamed or possibly just in the wrong place.

File: build/NuGetAdditionalFiles/Microsoft.NETCore.Compilers.props [hunk 9]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
+<Project ToolsVersion="14.0" DefaultTargets="Build" InitialTargets="MakeCompilerScriptsExecutable" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <!-- Always use the local build task, even if we just shell out to an exe in case there are
+       new properties in the local build task. -->
+  <UsingTask TaskName="Microsoft.CodeAnalysis.BuildTasks.Csc"
```
Comment (jaredpar) [inline build/NuGetAdditionalFiles/Microsoft.NETCore.Compilers.props]: More and more tasks are being added to our MSBuild DLL now.  When we get out of sync what happens?  Will builds which rely on the new tasks simply fail beacuse we haven't listed them heer?

File: src/Compilers/CSharp/CscCore/csc.deps.json [hunk 10]
```diff
@@ -0,0 +1,135 @@
+{
```
Comment (jaredpar) [inline src/Compilers/CSharp/CscCore/csc.deps.json]: Why is this file getting checked in?  Not familiar with it but at a glance looks like a lock file.

File: build/NuGetAdditionalFiles/Microsoft.NETCore.Compilers.props (hunk 9 above)
Comment (agocke) [inline build/NuGetAdditionalFiles/Microsoft.NETCore.Compilers.props]: What new task are you referring to? This is the same Csc task we've always had.

File: src/Compilers/CSharp/CscCore/csc.deps.json (hunk 10 above)
Comment (agocke) [inline src/Compilers/CSharp/CscCore/csc.deps.json]: This is the file that the shared framework host uses to resolve dependencies from the shared framework.

File: build/NuGetAdditionalFiles/Microsoft.NETCore.Compilers.props (hunk 8 above)
Comment (agocke) [inline build/NuGetAdditionalFiles/Microsoft.NETCore.Compilers.props]: Right -- this file is the equivalent of the Microsoft.Net.Compilers.props file. Why wouldn't it be in the same directory?

File: build/NuGetAdditionalFiles/Microsoft.NETCore.Compilers.props (hunk 9 above)
Comment (jaredpar) [inline build/NuGetAdditionalFiles/Microsoft.NETCore.Compilers.props]: Yep.  But we're getting more in the future: reference assembly helpers, editor config, etc ...  Will those need to be added here too?

File: src/Compilers/CSharp/CscCore/csc.deps.json (hunk 10 above)
Comment (jaredpar) [inline src/Compilers/CSharp/CscCore/csc.deps.json]: That file though looks completely machine read / written only.  Don't think any humans are going to be messing with that.  That makes me not sure how we manage this file going forward.

**Comment by agocke** [comment]
Closing because I'm readoing this change on top of the SDK changes.

**Pull request closed without merging**